### PR TITLE
Subscription: fix some issues on CN

### PR DESCRIPTION
**Pull request by DanielWang2035**
This PR:
- Takes pipe lock properly during creating and dropping subscriptions.
- Simplifies subscription procedures by merging multiple AlterTopicPlans and CreatePipePlanV2s into a single plan, and record multiple plan's failing index for partial rollback.
- Adds some UT
- subscription meta sync from CN to DN (currently no need for DN to push meta to CN)

This PR has:
- [x] been self-reviewed.

### CODE REVIEW

File: iotdb-core/confignode/src/main/java/org/apache/iotdb/confignode/procedure/impl/subscription/SubscriptionOperation.java [hunk 1]
```diff
@@ -28,6 +28,8 @@ public enum SubscriptionOperation {
   ALTER_CONSUMER_GROUP("alter consumer group"),
   CREATE_SUBSCRIPTION("create subscription"),
   DROP_SUBSCRIPTION("drop subscription"),
+  SYNC_CONSUMER_GROUP_META("syncConsumerGroupMeta"),
```
Comment (Caideyipi) [inline iotdb-core/confignode/src/main/java/org/apache/iotdb/confignode/procedure/impl/subscription/SubscriptionOperation.java:31]: The styles shall be equal

File: iotdb-core/confignode/src/main/java/org/apache/iotdb/confignode/persistence/pipe/PipeInfo.java [hunk 2]
```diff
@@ -160,6 +161,16 @@ public TSStatus alterPipe(AlterPipePlanV2 plan) {
     }
   }
 
+  public TSStatus createMultiplePipes(OperateMultiplePipesPlanV2 plans) {
```
Comment (Caideyipi) [inline iotdb-core/confignode/src/main/java/org/apache/iotdb/confignode/persistence/pipe/PipeInfo.java:164]: create?

File: iotdb-core/datanode/src/main/java/org/apache/iotdb/db/protocol/thrift/impl/DataNodeInternalRPCServiceImpl.java [hunk 3]
```diff
@@ -1062,11 +1136,81 @@ public TPushTopicMetaResp pushSingleTopicMeta(TPushSingleTopicMetaReq req) {
...
+      }
+
+      return hasException
+          ? new TPushTopicMetaResp()
+              .setStatus(new TSStatus(TSStatusCode.TOPIC_PUSH_META_ERROR.getStatusCode()))
+              .setExceptionMessages(exceptionMessages)
+          : new TPushTopicMetaResp()
+              .setStatus(new TSStatus(TSStatusCode.SUCCESS_STATUS.getStatusCode()));
+    } catch (Exception e) {
+      LOGGER.error("Error occurred when pushing multi topic meta", e);
```
Comment (SteveYurongSu) [inline iotdb-core/datanode/src/main/java/org/apache/iotdb/db/protocol/thrift/impl/DataNodeInternalRPCServiceImpl.java:1179]:
```suggestion
      LOGGER.warn("Error occurred when pushing multi topic meta", e);
```

File: iotdb-core/datanode/src/main/java/org/apache/iotdb/db/protocol/thrift/impl/DataNodeInternalRPCServiceImpl.java [hunk 4]
```diff
@@ -1062,11 +1136,81 @@ public TPushTopicMetaResp pushSingleTopicMeta(TPushSingleTopicMetaReq req) {
...
+          SubscriptionAgent.consumer().handleConsumerGroupMetaChanges(consumerGroupMetas);
+
+      return exceptionMessage == null
+          ? new TPushConsumerGroupMetaResp()
+              .setStatus(new TSStatus(TSStatusCode.SUCCESS_STATUS.getStatusCode()))
+          : new TPushConsumerGroupMetaResp()
+              .setStatus(new TSStatus(TSStatusCode.CONSUMER_PUSH_META_ERROR.getStatusCode()))
+              .setExceptionMessages(Collections.singletonList(exceptionMessage));
+    } catch (Exception e) {
+      LOGGER.error("Error occurred when pushing consumer group meta", e);
```
Comment (SteveYurongSu) [inline iotdb-core/datanode/src/main/java/org/apache/iotdb/db/protocol/thrift/impl/DataNodeInternalRPCServiceImpl.java:1203]:
```suggestion
      LOGGER.warn("Error occurred when pushing consumer group meta", e);
```

File: iotdb-core/datanode/src/main/java/org/apache/iotdb/db/protocol/thrift/impl/DataNodeInternalRPCServiceImpl.java [hunk 5]
```diff
@@ -1034,10 +1038,80 @@ public TPushPipeMetaResp pushSinglePipeMeta(TPushSinglePipeMetaReq req) {
...
+          SubscriptionAgent.topic().handleTopicMetaChanges(topicMetas);
+
+      return exceptionMessage == null
+          ? new TPushTopicMetaResp()
+              .setStatus(new TSStatus(TSStatusCode.SUCCESS_STATUS.getStatusCode()))
+          : new TPushTopicMetaResp()
+              .setStatus(new TSStatus(TSStatusCode.TOPIC_PUSH_META_ERROR.getStatusCode()))
+              .setExceptionMessages(Collections.singletonList(exceptionMessage));
+    } catch (Exception e) {
+      LOGGER.error("Error occurred when pushing topic meta", e);
```
Comment (SteveYurongSu) [inline iotdb-core/datanode/src/main/java/org/apache/iotdb/db/protocol/thrift/impl/DataNodeInternalRPCServiceImpl.java:1105]:
```suggestion
      LOGGER.warn("Error occurred when pushing topic meta", e);
```

File: iotdb-core/datanode/src/main/java/org/apache/iotdb/db/protocol/thrift/impl/DataNodeInternalRPCServiceImpl.java [hunk 6]
```diff
@@ -1034,10 +1038,80 @@ public TPushPipeMetaResp pushSinglePipeMeta(TPushSinglePipeMetaReq req) {
...
+      }
+
+      return hasException
+          ? new TPushPipeMetaResp()
+              .setStatus(new TSStatus(TSStatusCode.PIPE_PUSH_META_ERROR.getStatusCode()))
+              .setExceptionMessages(exceptionMessages)
+          : new TPushPipeMetaResp()
+              .setStatus(new TSStatus(TSStatusCode.SUCCESS_STATUS.getStatusCode()));
+    } catch (Exception e) {
+      LOGGER.error("Error occurred when pushing multi pipe meta", e);
```
Comment (SteveYurongSu) [inline iotdb-core/datanode/src/main/java/org/apache/iotdb/db/protocol/thrift/impl/DataNodeInternalRPCServiceImpl.java:1081]:
```suggestion
      LOGGER.warn("Error occurred when pushing multi pipe meta", e);
```

File: iotdb-core/confignode/src/main/java/org/apache/iotdb/confignode/procedure/impl/subscription/topic/runtime/TopicMetaSyncProcedure.java [hunk 7]
```diff
@@ -0,0 +1,160 @@
...
+          env.getConfigManager()
+              .getConsensusManager()
+              .write(new TopicHandleMetaChangePlan(topicMetaList));
+    } catch (ConsensusException e) {
+      LOGGER.warn("Failed in the write API executing the consensus layer due to: ", e);
+      response = new TSStatus(TSStatusCode.EXECUTE_STATEMENT_ERROR.getStatusCode());
+      response.setMessage(e.getMessage());
+    }
+    if (response.getCode() != TSStatusCode.SUCCESS_STATUS.getStatusCode()) {
+      throw new PipeException(response.getMessage());
```
Comment (SteveYurongSu) [inline iotdb-core/confignode/src/main/java/org/apache/iotdb/confignode/procedure/impl/subscription/topic/runtime/TopicMetaSyncProcedure.java:106]:
```suggestion
      throw new SubscriptionException(response.getMessage());
```

File: iotdb-core/confignode/src/main/java/org/apache/iotdb/confignode/procedure/impl/subscription/topic/runtime/TopicMetaSyncProcedure.java [hunk 8]
```diff
@@ -0,0 +1,160 @@
...
+    stream.writeShort(ProcedureType.TOPIC_META_SYNC_PROCEDURE.getTypeCode());
+    super.serialize(stream);
+  }
+
+  @Override
+  public boolean equals(Object o) {
+    if (this == o) {
+      return true;
+    }
+    return o instanceof TopicMetaSyncProcedure;
```
Comment (SteveYurongSu) [inline iotdb-core/confignode/src/main/java/org/apache/iotdb/confignode/procedure/impl/subscription/topic/runtime/TopicMetaSyncProcedure.java:153]:
```suggestion
    return o instanceof TopicMetaSyncProcedure;
```

File: iotdb-core/confignode/src/main/java/org/apache/iotdb/confignode/procedure/impl/subscription/topic/runtime/TopicMetaSyncProcedure.java [hunk 9]
```diff
@@ -0,0 +1,160 @@
...
+    }
+  }
+
+  @Override
+  public void executeFromOperateOnDataNodes(ConfigNodeProcedureEnv env) throws IOException {
+    LOGGER.info("TopicMetaSyncProcedure: executeFromOperateOnDataNodes");
+
+    Map<Integer, TPushTopicMetaResp> respMap = pushTopicMetaToDataNodes(env);
+    if (pushTopicMetaHasException(respMap)) {
+      throw new PipeException(
```
Comment (SteveYurongSu) [inline iotdb-core/confignode/src/main/java/org/apache/iotdb/confignode/procedure/impl/subscription/topic/runtime/TopicMetaSyncProcedure.java:116]:
```suggestion
      throw new SubscriptionException(
```

File: iotdb-core/confignode/src/main/java/org/apache/iotdb/confignode/procedure/impl/subscription/consumer/runtime/ConsumerGroupMetaSyncProcedure.java [hunk 10]
```diff
@@ -0,0 +1,161 @@
...
+    stream.writeShort(ProcedureType.CONSUMER_GROUP_META_SYNC_PROCEDURE.getTypeCode());
+    super.serialize(stream);
+  }
+
+  @Override
+  public boolean equals(Object o) {
+    if (this == o) {
+      return true;
+    }
+    return o instanceof ConsumerGroupMetaSyncProcedure;
```
Comment (SteveYurongSu) [inline iotdb-core/confignode/src/main/java/org/apache/iotdb/confignode/procedure/impl/subscription/consumer/runtime/ConsumerGroupMetaSyncProcedure.java:154]:
```suggestion
    return o instanceof ConsumerGroupMetaSyncProcedure;
```

File: iotdb-core/confignode/src/main/java/org/apache/iotdb/confignode/procedure/impl/subscription/consumer/runtime/ConsumerGroupMetaSyncProcedure.java [hunk 11]
```diff
@@ -0,0 +1,161 @@
...
+    }
+  }
+
+  @Override
+  public void executeFromOperateOnDataNodes(ConfigNodeProcedureEnv env) throws IOException {
+    LOGGER.info("ConsumerGroupMetaSyncProcedure: executeFromOperateOnDataNodes");
+
+    Map<Integer, TPushConsumerGroupMetaResp> respMap = pushConsumerGroupMetaToDataNodes(env);
+    if (pushConsumerGroupMetaHasException(respMap)) {
+      throw new PipeException(
```
Comment (SteveYurongSu) [inline iotdb-core/confignode/src/main/java/org/apache/iotdb/confignode/procedure/impl/subscription/consumer/runtime/ConsumerGroupMetaSyncProcedure.java:117]:
```suggestion
      throw new SubscriptionException(
```

File: iotdb-core/confignode/src/main/java/org/apache/iotdb/confignode/manager/subscription/SubscriptionMetaSyncer.java [hunk 12]
```diff
@@ -0,0 +1,118 @@
...
+
+  private final ConfigManager configManager;
+
+  private Future<?> metaSyncFuture;
+
+  // This variable is used to record whether the last sync operation was successful. If successful,
+  // before the next sync operation, it can be determined based on SubscriptionInfoVersion whether
+  // the
+  // metadata in DN and the latest metadata in CN have been synchronized, thereby skipping
+  // unnecessary sync operations.
```
Comment (SteveYurongSu) [inline iotdb-core/confignode/src/main/java/org/apache/iotdb/confignode/manager/subscription/SubscriptionMetaSyncer.java:58]: spotless

File: iotdb-core/confignode/src/main/java/org/apache/iotdb/confignode/procedure/env/ConfigNodeProcedureEnv.java [hunk 13]
```diff
@@ -726,6 +730,55 @@ public Map<Integer, TPushPipeMetaResp> dropSinglePipeOnDataNodes(String pipeName
...
+    final TPushMultiPipeMetaReq request =
+        new TPushMultiPipeMetaReq().setPipeMetas(pipeMetaBinaryList);
+
+    final AsyncClientHandler<TPushMultiPipeMetaReq, TPushPipeMetaResp> clientHandler =
+        new AsyncClientHandler<>(
+            DataNodeRequestType.PIPE_PUSH_MULTI_META, request, dataNodeLocationMap);
+    AsyncDataNodeClientPool.getInstance()
+        .sendAsyncRequestToDataNodeWithRetryAndTimeoutInMs(
+            clientHandler,
+            PipeConfig.getInstance().getPipeMetaSyncerSyncIntervalMinutes() * 60 * 1000 * 2);
```
Comment (SteveYurongSu) [inline iotdb-core/confignode/src/main/java/org/apache/iotdb/confignode/procedure/env/ConfigNodeProcedureEnv.java:746]:
```suggestion
            PipeConfig.getInstance().getPipeMetaSyncerSyncIntervalMinutes() * 60 * 1000 * 2 / 3);
```

File: iotdb-core/confignode/src/main/java/org/apache/iotdb/confignode/procedure/env/ConfigNodeProcedureEnv.java [hunk 14]
```diff
@@ -726,6 +730,55 @@ public Map<Integer, TPushPipeMetaResp> dropSinglePipeOnDataNodes(String pipeName
...
+        new TPushMultiPipeMetaReq().setPipeNamesToDrop(pipeNamesToDrop);
+
+    final AsyncClientHandler<TPushMultiPipeMetaReq, TPushPipeMetaResp> clientHandler =
+        new AsyncClientHandler<>(
+            DataNodeRequestType.PIPE_PUSH_MULTI_META, request, dataNodeLocationMap);
+    AsyncDataNodeClientPool.getInstance()
+        .sendAsyncRequestToDataNodeWithRetryAndTimeoutInMs(
+            clientHandler,
+            PipeConfig.getInstance().getPipeMetaSyncerSyncIntervalMinutes() * 60 * 1000 * 2);
+    return clientHandler.getResponseMap();
```
Comment (SteveYurongSu) [inline iotdb-core/confignode/src/main/java/org/apache/iotdb/confignode/procedure/env/ConfigNodeProcedureEnv.java:763]:
```suggestion
            PipeConfig.getInstance().getPipeMetaSyncerSyncIntervalMinutes() * 60 * 1000 * 2 / 3);
    return clientHandler.getResponseMap();
```

File: iotdb-core/confignode/src/main/java/org/apache/iotdb/confignode/procedure/env/ConfigNodeProcedureEnv.java [hunk 15]
```diff
@@ -755,6 +808,56 @@ public List<TSStatus> dropSingleTopicOnDataNode(String topicNameToDrop) {
...
+    final TPushMultiTopicMetaReq request =
+        new TPushMultiTopicMetaReq().setTopicMetas(topicMetaBinaryList);
+
+    final AsyncClientHandler<TPushMultiTopicMetaReq, TPushTopicMetaResp> clientHandler =
+        new AsyncClientHandler<>(
+            DataNodeRequestType.TOPIC_PUSH_MULTI_META, request, dataNodeLocationMap);
+    AsyncDataNodeClientPool.getInstance()
+        .sendAsyncRequestToDataNodeWithRetryAndTimeoutInMs(
+            clientHandler,
+            PipeConfig.getInstance().getPipeMetaSyncerSyncIntervalMinutes() * 60 * 1000 * 2);
```
Comment (SteveYurongSu) [inline iotdb-core/confignode/src/main/java/org/apache/iotdb/confignode/procedure/env/ConfigNodeProcedureEnv.java:824]:
```suggestion
            PipeConfig.getInstance().getPipeMetaSyncerSyncIntervalMinutes() * 60 * 1000 * 2 / 3);
```

File: iotdb-core/confignode/src/main/java/org/apache/iotdb/confignode/procedure/env/ConfigNodeProcedureEnv.java [hunk 16]
```diff
@@ -755,6 +808,56 @@ public List<TSStatus> dropSingleTopicOnDataNode(String topicNameToDrop) {
...
+    final TPushMultiTopicMetaReq request =
+        new TPushMultiTopicMetaReq().setTopicNamesToDrop(topicNamesToDrop);
+
+    final AsyncClientHandler<TPushMultiTopicMetaReq, TPushTopicMetaResp> clientHandler =
+        new AsyncClientHandler<>(
+            DataNodeRequestType.TOPIC_PUSH_MULTI_META, request, dataNodeLocationMap);
+    AsyncDataNodeClientPool.getInstance()
+        .sendAsyncRequestToDataNodeWithRetryAndTimeoutInMs(
+            clientHandler,
+            PipeConfig.getInstance().getPipeMetaSyncerSyncIntervalMinutes() * 60 * 1000 * 2);
```
Comment (SteveYurongSu) [inline iotdb-core/confignode/src/main/java/org/apache/iotdb/confignode/procedure/env/ConfigNodeProcedureEnv.java:840]:
```suggestion
            PipeConfig.getInstance().getPipeMetaSyncerSyncIntervalMinutes() * 60 * 1000 * 2 / 3);
```

Comment (SteveYurongSu) [reply iotdb-core/confignode/src/main/java/org/apache/iotdb/confignode/procedure/env/ConfigNodeProcedureEnv.java:840]: check others implementation

File: iotdb-core/confignode/src/main/java/org/apache/iotdb/confignode/manager/subscription/SubscriptionMetaSyncer.java [hunk 17]
```diff
@@ -0,0 +1,118 @@
...
+  private boolean isLastSubscriptionSyncSuccessful = false;
+
+  SubscriptionMetaSyncer(ConfigManager configManager) {
```
Comment (SteveYurongSu) [inline iotdb-core/confignode/src/main/java/org/apache/iotdb/confignode/manager/subscription/SubscriptionMetaSyncer.java:61]: TODO: enable

**Review by SteveYurongSu — APPROVED** [review]

**Pull request merged**